### PR TITLE
Check out repository as user

### PR DESCRIPTION
**Pull request by hp197**
This way all checked out files belong to the user and not root.

Signed-off-by: Henry Pauli <henry@mixict.nl>

### CODE REVIEW

**Comment by lbetz** [comment]
Fine, but why? Puppet has to run as root, modules directory is owned by root.

**Comment by hp197** [comment]
But my webserver isn't running as root, it is running as the user apache.
So I want the files in the module to have the same access rights as the rest (and as the running php-fpm user)

**Comment by lbetz** [comment]
Why should the web server have to?

**Comment by hp197** [comment]
Because currently, vcsrepo checks out the repo as puppet user (root:root) with chmod 0750.
Apache/php are unable to access those files, as they run under their own user (and group).

Note that I dont have a docker setup, where everything runs as root (/one) user.

**Comment by lbetz** [comment]
Giving the modules to the user of the web server as owner is a greater risk than giving your root user an umask of 022 for the puppet run.

If you really wanna do this, use a default resource of vcsrepo with $conf_user in correct scope in your own code.

**Comment by hp197** [comment]
Sorry, I fail to see how giving the ownership of the modules to the user running the webserver is insecure.

A umask of 022 is insecure and not recommended by different security baselines, for example well known industry standard CIS baseline:
https://www.itsecure.hu/library/image/CIS_Red_Hat_Enterprise_Linux_7_Benchmark_v2.1.1.pdf
> Page 294
> 5.4.4 Ensure default user umask is 027 or more restrictive (Scored)

**Comment by lbetz** [comment]
I said more insecure. The problem in general is the use of git when a package is more useful. And if you want it then do it yourself as described above.

**Comment by widhalmt** [comment]
You definitely don't want your webserver user own the files your webserver is serving. If someone is cracking up your webserver, they can change the files just as they please. Your webserver needs to be able to *read* the files, not write, let alone own them.

And yes, a umask of 022 for root might not be a good idea after all but we're not talking about changing it as a whole, only for the files this very puppet module is writing.

In short: Files owned by a user that might not be easily tampered with but readable by the webserver is the way to go.

By the way, that's true even aside from puppet, you might want to recheck the ownerships and permissions of the files in your web root. Only files that have to be written by the webserver (uploads, cache) should be owned by the webserver user. Hint: self-updating CMS might be *convenient* but they are not *secure*.

**Comment by hp197** [comment]
I still dont agree that it is inherently unsafe that the that the user runs the web/php server is should never has write access to the files. What do you think shared webhosting companies do? You write your files under your user, in your own homedir and they run a php-fpm as your very own user.

Anyway, the opinion of the maintainers is clearly different so i'll close the PR

**Comment by Thomas-Gelf** [comment]
@hp197: I wouldn't consider shared web hosting companies the gold standard when it comes to securing your application :rofl: They try to protect themselves from your application and they try to protect users from each other. But they do not see it as their job to protect your application. That's what makes it so easy for intruders to install back-doors as soon as they spot an attack vector.

The main problem here seems to be that a non-default umask renders Puppet Runs unpredictable. This has been [reported](https://tickets.puppetlabs.com/browse/PUP-2359) long time ago. Your daemons hopefully do not run with a custom umask, but your manual `puppet agent -t` run does. And while there is a way to set sane defaults for File resources, this doesn't work for `vcsrepo`.

So I'd strongly suggest using `022` when running Puppet.

Cheers,
Thomas
